bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/HeapSizeManager.java

```diff
@@ -63,61 +76,77 @@ public long getMaxHeapSize() {
     return maxHeapSize;
   }
 
-  public synchronized void waitUntilAllOperationsAreDone() throws InterruptedException {
+  public synchronized void flush() throws InterruptedException {
+    isFlushing.set(true);
     boolean performedWarning = false;
-    while(!pendingOperationsWithSize.isEmpty()) {
-      cleanupFinishedOperations();
-      if (pendingOperationsWithSize.isEmpty()) {
-        break;
-      }
-      if (!performedWarning
-          && lastOperationChange + INTERVAL_NO_SUCCESS_WARNING < System.currentTimeMillis()) {
-        long lastUpdated = (System.currentTimeMillis() - lastOperationChange) / 1000;
-        LOG.warn("No operations completed within the last %d seconds."
-            + "There are still %d operations in progress.", lastUpdated,
-          pendingOperationsWithSize.size());
+    cleanupFinishedOperations();
+    while (!pendingOperationsWithSize.isEmpty()) {
+      if (!performedWarning && requiresWarning()) {
+        warn();
         performedWarning = true;
-        waitForCompletions(FINISH_WAIT_MILLIS);
       }
+      wait(FINISH_WAIT_MILLIS);
+      cleanupFinishedOperations();
     }
     if (performedWarning) {
-      LOG.info("flush() completed");
+      LOG.info("flush() completed. See the log for previous warnings.");
     }
+    isFlushing.set(false);
+  }
+
+  /**
+   * @return {@code true} if the last successful completion is awhile ago (longer than {@code
+   *         INTERVAL_NO_SUCCESS_WARNING_MS} ms ago.
+   */
+  private boolean requiresWarning() {
+    return lastOperationChange + INTERVAL_NO_SUCCESS_WARNING_MS < System.currentTimeMillis();
```

Comment:
lastOperationChange doesn't represent last successful completion because we also update this ts on register.